File: Game.java
import java.util.*;
import java.io.*;
import java.math.*;

class Point {
    int x, y;

    public Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public String toString() {
        return String.format("(x=%d, y=%d)", x, y);
    }

    public boolean equals(Point other) {
        return this.x == other.x && this.y == other.y;
    }
}

class Unit {
    Point position;
    char level;

    static HashMap<String, Point> actionToPos; // action -> (x, y)
    static HashMap<String, String[]> pushToDirections;

    static {
        actionToPos = new HashMap<String, Point>();
        actionToPos.put("NW",   new Point(-1, -1));
        actionToPos.put("N",    new Point(0, -1));
        actionToPos.put("NE",   new Point(1, -1));
        actionToPos.put("W",    new Point(-1, 0));
        actionToPos.put("E",    new Point(1, 0));
        actionToPos.put("SW",   new Point(-1, 1));
        actionToPos.put("S",    new Point(0, 1));
        actionToPos.put("SE",   new Point(1, 1));

        pushToDirections = new HashMap<String, String[]>();
        pushToDirections.put("NW",  new String[]{"N", "NW", "W"});
        pushToDirections.put("N",  new String[]{"N", "NW", "NE"});
        pushToDirections.put("NE",  new String[]{"N", "NE", "E"});
        pushToDirections.put("W",  new String[]{"NW", "SW", "W"});
        pushToDirections.put("E",  new String[]{"NE", "SE", "E"});
        pushToDirections.put("SW",  new String[]{"W", "SW", "S"});
        pushToDirections.put("S",  new String[]{"SW", "S", "SE"});
        pushToDirections.put("SE",  new String[]{"S", "SE", "E"});
    }

    public Unit(int x, int y) {
        this.position = new Point(x, y);
    }

    public Unit(int x, int y, char[][] map) {
        this(x, y);
        this.level = map[y][x];
    }

    public Unit(Unit other) {
        this(other.position.x, other.position.y);
        this.level = other.level;
    }

    public String toString() {
        return position.toString();
    }

    public void moveAndBuild(Action action, char[][] map) {
        move(action.move);
        build(action.build, map);
    }

    public void pushAndBuild(Action action, Unit[] pushedUnits, char[][] map) {
        push(action.move, action.build, pushedUnits);
        build(action.move, map);
    }

    public void push(String push, String where, Unit[] pushedUnits) {
        Point pushDirection = actionToPos.get(push);
        int x = position.x + pushDirection.x;
        int y = position.y + pushDirection.y;

        for (Unit unit : pushedUnits) {
            if (unit.position.x == x && unit.position.y == y) {
                // we found unit to be pushed
                Point whereDirection = actionToPos.get(where);
                unit.position.x += whereDirection.x;
                unit.position.y += whereDirection.y;
                return;
            }
        }

        throw new RuntimeException("SHOULD NOT BE REACHED!!! DONT MAKE BUGS PLS");
    }

    public void move(String move) {
        Point direction = actionToPos.get(move);
        position.x += direction.x;
        position.y += direction.y;
    }

    public void build(String build, char[][] map) {
        Point direction = actionToPos.get(build);
        map[position.y + direction.y][position.x + direction.x]++;
    }
}

class Action {
    String atype;
    int index;
    String move;
    String build;

    // when action executed
    Point position;
    int level;

    public Action(String atype, int index, String move, String build) {
        this.atype = atype;
        this.index = index;
        this.move = move;
        this.build = build;
    }

    public String toString() {
        return String.format("%s %d %s %s", atype, index, move, build);
    }
}

class State {
    static int size;
    static int unitsPerGame;

    char[][] map;

    Unit[] myUnits;
    Unit[] enemyUnits;

    boolean myUnitTurn;

    int myScore, enemyScore;

    static ArrayList<String> directions;

    static {
        directions = new ArrayList<String>();
        directions.add("NW");
        directions.add("N");
        directions.add("NE");
        directions.add("W");
        directions.add("E");
        directions.add("SW");
        directions.add("S");
        directions.add("SE");
    }

    public State(boolean myUnitTurn) {
        map = new char[size][size];
        this.myUnitTurn = myUnitTurn;
    }

    public void readState(Scanner input) {
        for (int i = 0; i < size; i++) {
            String row = input.next();
            map[i] = row.toCharArray();
        }

        myUnits = new Unit[unitsPerGame];
        for (int i = 0; i < unitsPerGame; i++) {
            myUnits[i] = new Unit(input.nextInt(), input.nextInt(), map);
        }
        enemyUnits = new Unit[unitsPerGame];
        for (int i = 0; i < unitsPerGame; i++) {
            enemyUnits[i] = new Unit(input.nextInt(), input.nextInt(), map);
        }
        int legalActions = input.nextInt();
        for (int i = 0; i < legalActions; i++) {
            Action a = new Action(
                input.next(), // atype
                input.nextInt(), // index
                input.next(), // move
                input.next()
            ); // build
            System.err.println(a);
        }
    }

    public void printMap() {
        System.err.print(" ");
        for (int i = 0; i < map.length; i++) {
            System.err.print(i);
        }
        System.err.println();
        for (int i = 0; i < map.length; i++) {
            System.err.print(i);
            for (int j = 0; j < map[i].length; j++) {
                System.err.print(map[i][j]);
            }
            System.err.println();
        }
        for (int i = 0; i < unitsPerGame; i++) {
            System.err.println("My unit: " + myUnits[i]);
            System.err.println("Enemy unit: " + enemyUnits[i]);
        }
    }

    public State copyState(boolean switchTurn) {
        State copied = new State(this.myUnitTurn);
        copied.map = new char[size][size];
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                copied.map[i][j] = map[i][j];
            }
        }

        copied.myUnits = new Unit[unitsPerGame];
        copied.enemyUnits = new Unit[unitsPerGame];
        for (int i = 0; i < unitsPerGame; i++) {
            copied.myUnits[i] = new Unit(myUnits[i]);
            copied.enemyUnits[i] = new Unit(enemyUnits[i]);
        }
        copied.myUnitTurn = switchTurn ? !myUnitTurn : myUnitTurn;

        copied.myScore = myScore;
        copied.enemyScore = enemyScore;

        return copied;
    }

    public State executeAction(Action action) {
        Unit executingUnit = myUnitTurn ? myUnits[action.index] : enemyUnits[action.index];
        switch (action.atype) {
            case "MOVE&BUILD":
                executingUnit.moveAndBuild(action, map);
                if (map[executingUnit.position.y][executingUnit.position.x] == '3') {
                    if (myUnitTurn) {
                        myScore++;
                    } else {
                        enemyScore++;
                    }
                }
                break;
            case "PUSH&BUILD":
                Unit[] otherUnits = myUnitTurn ? enemyUnits : myUnits;
                executingUnit.pushAndBuild(action, otherUnits, map);
                break;
        }
        return this;
    }

    public Point canMove(String move, Point position, Unit[] units, Unit[] otherUnits, char[][] map) {
        Point moveTo = canBuild(move, position, units, otherUnits, map);
        if (moveTo == null) {
            return null;
        }

        char currentLevel = map[position.y][position.x];
        char moveToLevel = map[moveTo.y][moveTo.x];
        if (currentLevel + 1 < moveToLevel) {
            return null;
        }

        return moveTo;
    }

    public Point canBuild(String move, Point position, Unit[] units, Unit[] otherUnits, char[][] map) {
        Point direction = Unit.actionToPos.get(move);
        int x = position.x + direction.x;
        int y = position.y + direction.y;

        if (!legalCell(x, y, map)) {
            return null;
        }

        for (Unit unit : units) {
            if (unit.position.x == x && unit.position.y == y) {
                return null;
            }
        }
        for (Unit unit : otherUnits) {
            if (unit.position.x == x && unit.position.y == y) {
                return null;
            }
        }
        return new Point(x, y);
    }

    public boolean legalCell(int x, int y, char[][] map) {
        if (x < 0 || y < 0 || x >= State.size || y >= State.size) {
            return false;
        }
        if (map[y][x] == '.' || map[y][x] == '4') {
            return false;
        }

        return true;
    }

    public Unit getNeighbourUnit(Unit myUnit, String direction, Unit[] units, char[][] map) {
        Point directionPoint = Unit.actionToPos.get(direction);
        int x = myUnit.position.x + directionPoint.x;
        int y = myUnit.position.y + directionPoint.y;

        if (!legalCell(x, y, map)) {
            return null;
        }

        for (Unit unit : units) {
            if (unit.position.x == x && unit.position.y == y) {
                return unit;
            }
        }

        return null;
    }

    // public Unit canPush(String push, Point position, Unit[] units, Unit[] otherUnits, char[][] map) {
        // Point direction = Unit.actionToPos.get(push);
        // int x = position.x + direction.x;
        // int y = position.y + direction.y;

        // if (!legalCell(x, y, map)) {
            // return null;
        // }

        // for (Unit unit : otherUnits) {
            // if (unit.position.x == x && unit.position.y == y) {
                // int pushedX = x + direction.x;
                // int pushedY = y + direction.y;
                // if (legalCell(pushedX, pushedY, map)) {
                    // return unit;
                // }
                // return null;
            // }
        // }
        // return null;
    // }

    public ArrayList<Action> legalActions() {
        ArrayList<Action> legalActions = new ArrayList<Action>();
        Unit[] units = myUnitTurn ? myUnits : enemyUnits;
        Unit[] otherUnits = myUnitTurn ? enemyUnits : myUnits;

        for (int i = 0; i < unitsPerGame; i++) {
            Unit unit = units[i];
            // MOVE&BUILD
            // first move
            for (String moveDirection : directions) {
                Point movePoint = canMove(moveDirection, unit.position, units, otherUnits, map);
                if (movePoint != null) {
                    // then build
                    int tmpX = unit.position.x;
                    int tmpY = unit.position.y;
                    unit.position.x = -1;
                    unit.position.y = -1;
                    for (String buildDirection : directions) {
                        Point buildPoint = canBuild(buildDirection, movePoint, units, otherUnits, map);
                        if (buildPoint != null) {
                            Action action = new Action("MOVE&BUILD", i, moveDirection, buildDirection);
                            legalActions.add(action);
                        }
                    }
                    unit.position.x = tmpX;
                    unit.position.y = tmpY;
                }
            }

            // PUSH&BUILD
            // push
            for (String pushDirection : directions) {
                Unit pushedUnit = getNeighbourUnit(unit, pushDirection, otherUnits, map);
                if (pushedUnit != null) {
                    // System.err.println("mozem tiskat " + pushDirection);
                    for (String moveDirection : Unit.pushToDirections.get(pushDirection)) {
                        // check whether pushedTo position is free
                        Point pushedTo = canMove(moveDirection, pushedUnit.position, units, otherUnits, map);
                        if (pushedTo != null) {
                            Action action = new Action("PUSH&BUILD", i, pushDirection, moveDirection);
                            legalActions.add(action);
                        }
                    }
                }
            }
        }
        return legalActions;
    }
}

class Game {

    Scanner input;

    State state;

    Random random;

    public Game() {
        input = new Scanner(System.in);

        State.size = input.nextInt();
        State.unitsPerGame = input.nextInt();

        state = new State(true);

        random = new Random();
    }

    public boolean checkPosition(Point position) {
        int x = position.x;
        int y = position.y;

        if (x < 0 || y < 0 || x >= State.size || y >= State.size) {
            return false;
        }
        // -2 == '.'
        return state.map[y][x] != -2 && state.map[y][x] != 4 && state.map[y][x] <= 10;
    }

    // public Action findMove() {
        // Action bestAction = new Action("MOVE&BUILD", 0, null, null);
        // for (String direction : directions) {
            // // System.err.println(action);
            // Point nextPosition = next(state.myUnits[0].position, direction);

            // if (!checkPosition(nextPosition)) {
                // continue;
            // }
            // int nextPositionLevel = state.map[nextPosition.y][nextPosition.x];
            // if (nextPositionLevel == state.myUnits[0].level + 1) {
                // bestAction.move = direction;
                // bestAction.level = nextPositionLevel;
                // bestAction.position = nextPosition;
                // // System.err.println(nextPosition);
                // return bestAction;
            // }
            // if (nextPositionLevel >= bestAction.level && nextPositionLevel <= state.myUnits[0].level + 1) {
                // bestAction.move = direction;
                // bestAction.level = nextPositionLevel;
                // bestAction.position = nextPosition;
            // }

        // }

        // if (bestAction.move == null) {
            // System.err.println("No possible action found");
        // }

        // return bestAction;
    // }

    // public Action findBuild(Action action) {
        // Point nextPosition = action.position;
        // int maxLevel = 0;
        // for (String direction : directions) {
            // Point buildPosition = next(nextPosition, direction);
            // if (!checkPosition(buildPosition)) {
                // continue;
            // }
            // int buildPositionLevel = state.map[buildPosition.y][buildPosition.x];
            // if (buildPositionLevel >= maxLevel && buildPositionLevel <= action.level) {
                // maxLevel = buildPositionLevel;
                // action.build = direction;
            // }
        // }
        // return action;
    // }

    public void mainLoop() {
        state.readState(input);
        state.printMap();
        while (true) {
            long time = System.currentTimeMillis();
            System.err.println(state.legalActions());
            System.err.println("passed time for legal actions :" + (System.currentTimeMillis() - time));
            Action a = new Action(
                input.next(), // atype
                input.nextInt(), // index
                input.next(), // move
                input.next()
            ); // build
            state.executeAction(a);
            state.printMap();
            System.err.println("SCORE: " + state.myScore);
        }
        // while (true) {
            // state.readState(input);
            // state.printMap();
            // Action action = findMove();
            // action = findBuild(action);
            // System.out.println(action);
        // }
    }

    public static void main(String args[]) {
        Game Game = new Game();
        Game.mainLoop();
    }
}

File: game.py
from random import randint, shuffle
from collections import namedtuple
from copy import deepcopy

DIRECTIONS = [(-1, -1),
              (-1, 0),
              (-1, 1),
              (0, -1),
              (0, 1),
              (1, -1),
              (1, 0),
              (1, 1)]

PUSH_DIRECTIONS = {(-1, -1): [(-1, -1),
                              (-1, 0),
                              (0, -1)],

                   (-1, 0): [(-1, -1),
                             (-1, 0),
                             (-1, 1)],

                   (-1, 1): [(-1, 0),
                             (-1, 1),
                             (0, 1)],

                   (0, -1): [(-1, -1),
                             (0, -1),
                             (1, -1)],

                   (0, 1): [(-1, 1),
                            (0, 1),
                            (1, 1)],

                   (1, -1): [(0, -1),
                             (1, -1),
                             (1, 0)],

                   (1, 0): [(1, -1),
                            (1, 0),
                            (1, 1)],

                   (1, 1): [(1, 0),
                            (1, 1),
                            (0, 1)]}

WORLD_SIZE = 5


class State():

    def __init__(self):
        self.world = [[0 for i in range(WORLD_SIZE)] for j in range(WORLD_SIZE)]

        positions = list([i, j] for i in range(WORLD_SIZE) for j in range(WORLD_SIZE))
        shuffle(positions)
        self.p1_units = positions[0:2]
        self.p2_units = positions[2:4]

        self.p1_score = 0
        self.p2_score = 0
        # self.my_turn = True if randint(0, 1) else False
        self.my_turn = True

    def __str__(self):
        result = ''
        for row in self.world:
            for char in row:
                result += str(char)
            result += '\n'
        result += '\n'

        with_positions = [[0 for i in range(WORLD_SIZE)] for j in range(WORLD_SIZE)]
        for unit in self.p1_units:
            with_positions[unit[1]][unit[0]] = 1
        for unit in self.p2_units:
            with_positions[unit[1]][unit[0]] = 2

        for row in with_positions:
            for char in row:
                result += str(char)
            result += '\n'
        result += '\n'
        result += str(self.p1_units) + '\n'
        result += str(self.p2_units) + '\n'
        result += str(self.p1_score) + '\n'
        result += str(self.p2_score) + '\n'
        return result


class Action():

    def __init__(self, atype, index, dir1, dir2, state):
        def get_level(pos, state):
            return state.world[pos[1]][pos[0]]

        self.type = atype
        self.index = index
        self.dir1 = dir1
        self.dir2 = dir2

        my_units = state.p1_units if state.my_turn else state.p2_units
        enemy_units = state.p2_units if state.my_turn else state.p1_units

        self.move_from = my_units[index]

        if atype == 'MOVE&BUILD':
            self.move_to = [self.move_from[0] + dir1[0], self.move_from[1] + dir1[1]]
            self.build_to = [self.move_to[0] + dir2[0], self.move_to[1] + dir2[1]]
            self.move_to_level = get_level(self.move_to, state)
        else:

            self.push_from = [self.move_from[0] + dir1[0], self.move_from[1] + dir1[1]]
            self.push_to = [self.push_from[0] + dir2[0], self.push_from[1] + dir2[1]]
            self.build_to = self.push_from
            self.push_from_level = get_level(self.push_from, state)
            self.push_to_level = get_level(self.push_to, state)

        self.move_from_level = get_level(self.move_from, state)
        self.build_level_before = get_level(self.build_to, state)

        self.state_before = state
        self.state_after = deepcopy(state)
        self.state_after = execute(self.state_after, self)



    def __repr__(self):
        return '{} {} {} {} '.format(self.type,
                                     self.index,
                                     self.dir1,
                                     self.dir2)


def can_move(unit, direction, units1, units2, world):
    moved_pos = can_build(unit, direction, units1, units2, world)
    if moved_pos is None:
        return None
    [old_x, old_y] = unit
    [x, y] = moved_pos
    current_height = world[old_y][old_x]
    new_height = world[y][x]
    if current_height + 1 < new_height:
        return None

    return moved_pos


def can_build(unit, direction, units1, units2, world):
    x = unit[0] + direction[0]
    y = unit[1] + direction[1]

    if not legal_cell(x, y, world):
        return None

    for u in units1:
        if [x, y] == u:
            return None

    for u in units2:
        if [x, y] == u:
            return None

    return [x, y]

def legal_cell(x, y, world):
    if (x < 0 or y < 0 or x >= WORLD_SIZE or y >= WORLD_SIZE):
        return False

    if (world[y][x] == -1 or world[y][x] == 4):
        return False

    return True

def legal_actions(state):
    actions = []
    my_units = state.p1_units if state.my_turn else state.p2_units
    enemy_units = state.p2_units if state.my_turn else state.p1_units

    for i in range(2):
        my_unit = my_units[i]
        for direction in DIRECTIONS:
            # MOVE&BUILD
            moved_pos = can_move(my_unit, direction, my_units, enemy_units, state.world)
            # print(my_unit, direction, moved_pos)
            if moved_pos is not None:
                tmp = my_unit[:]
                my_unit[0], my_unit[1] = -1, -1
                for build_dir in DIRECTIONS:
                    build_pos = can_build(moved_pos, build_dir, my_units, enemy_units, state.world)
                    if build_pos is not None:
                        action = Action('MOVE&BUILD', i, direction, build_dir, state)
                        actions.append(action)

                my_unit[0], my_unit[1] = tmp

            # PUSH&BUILD
            pushed_unit = get_neighbour_unit(my_unit, direction, enemy_units, state.world)
            if pushed_unit is not None:
                for push_dir in PUSH_DIRECTIONS[direction]:
                    pushed_to = can_move(pushed_unit, push_dir, my_units, enemy_units, state.world)
                    if pushed_to is not None:
                        action = Action('PUSH&BUILD', i, direction, push_dir, state)
                        actions.append(action)

    return actions


def get_neighbour_unit(my_unit, direction, enemy_units, world):
    x = my_unit[0] + direction[0]
    y = my_unit[1] + direction[1]

    if not legal_cell(x, y, world):
        return None

    for unit in enemy_units:
        if unit == [x, y]:
            return unit

    return None


def execute(state, action):
    unit = state.p1_units[action.index] if state.my_turn else state.p2_units[action.index]
    if action.type == 'MOVE&BUILD':
        unit[0] += action.dir1[0]
        unit[1] += action.dir1[1]
        x, y = (unit[0] + action.dir2[0], unit[1] + action.dir2[1])
        state.world[y][x] += 1
        if state.world[y][x] == 3:
            if state.my_turn:
                state.p1_score += 1
            else:
                state.p2_score += 1
    else:
        other_units = state.p2_units if state.my_turn else state.p1_units
        pushed_unit = [unit[0] + action.dir1[0], unit[1] + action.dir1[1]]

        for other in other_units:
            if other == pushed_unit:
                state.world[other[1]][other[0]] += 1
                other[0] += action.dir2[0]
                other[1] += action.dir2[1]
                break

    state.my_turn = not state.my_turn
    return state


if __name__ == "__main__":
    state = State()
    while True:
        actions = legal_actions(state)
        for idx, action in enumerate(actions):
            print(idx, action)
        print(state)
        user = int(input())
        state = execute(state, actions[user])
